cpp/src/arrow/type.h

```diff
@@ -147,6 +147,15 @@ struct Type {
   };
 };
 
+struct ARROW_EXPORT DataTypeLayout {
+  // The bit width for each buffer in this DataType's representation
+  // (-1 if the item size for a given buffer is unknown or variable,
+  //  0 if the buffer is always null).
+  // Child types are not included, they should be inspected separately.
+  std::vector<int64_t> bit_widths;
+  bool has_dictionary;
+};
```

Comment:
Why should children be excluded? It's also missing the union mode which dictates how to interpret the children arrays.

Reply:
For several reasons:
1. the caller may not be interested directly in children
2. handling the information for children would make the structure self-referential
3. the caller may want to do other things while recursing, or recurse selectively, etc.

As for the union mode, this struct only describes the physical layout, not how to interpret arrays. How to interpret arrays is embodied in the DataType itself.